Allow batch action from another controller

mpoiriert opened the pull request:
<!-- THE PR TEMPLATE IS NOT AN OPTION. DO NOT DELETE IT, MAKE SURE YOU READ AND EDIT IT! -->
## Allow batch action from another controller

<!-- Describe your Pull Request content here -->

<!--
    Show us you choose the right branch.
    Different branches are used for different things :
    - 4.x is for everything backwards compatible, like patches, features and deprecation notices
    - 5.x is for deprecation removals and other changes that cannot be done without a BC-break
    More details here: https://github.com/sonata-project/SonataAdminBundle/blob/4.x/CONTRIBUTING.md#base-branch
-->
I am targeting this branch, because I am adding a feature that is backward compatible.

<!--
    Specify which issues will be fixed/closed.
    Remove it if this is not related.
-->


## Changelog

<!-- REMOVE EMPTY SECTIONS -->
```markdown
### Added
- Added support to batch action via another controller with 'controller' configuration

### Deprecated
- batchAction{actionName}IsRelevant will be remove in version 5.0. Move logic to your action.
```

<!--
    If this is a work in progress, uncomment the "To do" section.
    You can add as many tasks as you want.
    If some are not relevant, just remove them.
-->
<!--
## To do

- [x] Update the tests;
- [x] Update the documentation;
- [ ] Add an upgrade note.
-->

mpoiriert committed:
Allow batch action from another controller

mpoiriert commented:
@VincentLanglet This is the PR for the batch action support in another controller.

I have tested it locally. I still need to do automation test and update the documentation but I want to make sure the way I have implemented is is accepted before doing any of these.

VincentLanglet commented:
I also ping @jordisala1991 
and @franmomu since he implemented the adminFetcher.

VincentLanglet commented on src/Controller/CRUDController.php:
```diff
@@ -439,12 +442,15 @@ public function batchAction(Request $request): Response
+
         $isRelevantAction = sprintf('batchAction%sIsRelevant', $camelizedAction);
 
         if (method_exists($this, $isRelevantAction)) {
```
You forgot to support the `isRelevant` check for your "custom controller".

mpoiriert replied:
I have a few bundle of sonata installed and I didn't see any place it have been implemented.

Is it use somewhere ? I think it's complicating something that could easily be done directly in the controller:


```PHP
class CustomController extends CRUDController
{
  public function myCustomAction($query, $request)
  {
      // logic to test if action is relevent
      if(!$this->isRelevent($query)) {
            $this->addFlash(
                'sonata_flash_info',
                $this->trans($nonRelevantMessage, [], 'SonataAdminBundle')
            );
            
            return $this->redirectToList();
      }
      
      // continue logic
  }
}
```

I have extended the CRUDController just to be consistent with the method call, I think we should extract some method method from the CRUD controller and put them in a trait that could be use for both CRUDController and custom action. It's already an issue with some of the function that could be useful in Admin related action (like renderWithExtraParams, isXmlHttpRequest, handleModelManagerThrowable, etc.)

VincentLanglet replied:
Dunno, never used it too. I wonder if we should deprecate (in another pr) this isRelevant feature.

mpoiriert replied:
the $this->isRevelant() dooes not exists as is, it was to show how to put it in the controller instead of a hook.

But yes we should probably deprecate the hook call so dev move it to their controller. Less logic to maintain on sonata side.

VincentLanglet commented on src/Controller/CRUDController.php:
```diff
@@ -524,7 +523,60 @@ public function batchAction(Request $request): Response
+            $request = new Request();
+            $request->attributes->set('_controller', $controller);
+            try {
+                return false !== $this->container->get('controller_resolver')->getController($request);
```
Rather than working with the controllerResolver, what do you think about creating something like a BatchActionResolver instead ? Something based on https://symfony.com/doc/current/service_container/tags.html

We could have
```
interface BatchActionInterface
{
    public function supports(AdminInterface $admin, string $action): bool;

    public function isRelevant(AdminInterface $admin, array $idx, bool $allElements, Request $request): string|bool;

    public function execute(AdminInterface $admin, ProxyQueryInterface $proxyQuery, Request $request): Response
}
```
and
```
class BatchActionResolver
{
    /** @var array<BatchActionInterface> $/
    private array $batchActions;

    public function __construct(array $batchActions)
    {
         $this->batchActions = $batchActions;
    }

    public function get(AdminInterface $adminInterface, string $action): BatchActionInterface
    {
        foreach ($this->batchActions as $batchAction) {
             if ($batchAction->supports($admin, $action) {
                 return $batchAction;
             }
        }
    }
}
```

And then the controller would do
```
$this->container->get('batch_action_resolver')->get($this->admin, $action);
```
And you don't even need to add a new `controller` key.

mpoiriert replied:
From my point of view this is overcomplexify something that is a concept already know by Symfony developer (create a controller) and will also support any feature related to a controller. It would need to test but service injection in controller method for example.

Also the BatchActionResorlver and BatchActionInterface the way it have been implement would limit one action per class. This might be something that people wish but Using the controller approach would give the choice to people to do what they want.

It will also load a bunch of service and executing some logic to find one (that need to be implemented) on top of having to configure priority (what if multiple action support it ?).

And the controller approach is easy to use directly with the AdminExtensionInterface, nothing to add there.

Also it's a lot easier to document. It could be added to the current documentation with a couple of lines.

mpoiriert replied:
I would add that people might then to extends the class from AbstractController or event CRUDController for some reason, they are most likely register it as a controller anyway.

VincentLanglet replied:
> From my point of view this is overcomplexify something that is a concept already know by Symfony developer (create a controller) and will also support any feature related to a controller. It would need to test but service injection in controller method for example.
> 
> Also the BatchActionResorlver and BatchActionInterface the way it have been implement would limit one action per class. This might be something that people wish but Using the controller approach would give the choice to people to do what they want.
> 
> It will also load a bunch of service and executing some logic to find one (that need to be implemented) on top of having to configure priority (what if multiple action support it ?).
> 
> And the controller approach is easy to use directly with the AdminExtensionInterface, nothing to add there.
> 
> Also it's a lot easier to document. It could be added to the current documentation with a couple of lines.

I wanted to have an interface to avoid relying on the name like it was before
```
$finalAction = sprintf('batchAction%s', $camelizedAction);
```
but I'm not familiar with the `_controller` param. Do you mean you're passing the controller method and not only the controller name ?
```
'_controller' => 'App\Controller\AdminController::batchClone',
```
and not 
```
'_controller' => 'App\Controller\AdminController',
```
it could be enough then.

Then. rather than your if/else strategy, can't we do
```
$camelizedAction = InflectorFactory::create()->build()->classify($action);
$finalAction = sprintf('batchAction%s', $camelizedAction);

$controller = $batchActions[$action]['controller'] ?? sprintf('%s::%s', self::class, $finalAction);

$request = new Request();
$request->attributes->set('_controller', $controller);
try {
    return false !== $this->container->get('controller_resolver')->getController($request);
} catch (\Throwable $error) {
     return false;
}
```
dunno if it's a good idea.

Also, in case they are extra values in the request, should we prefer `$request = clone $existingRequest` ?

VincentLanglet replied:
We might also want to deprecate the magic behind the `batchAction%s` name ; and prefer people to pass 
```
'_controller' => 'App\Controller\CRUDController::batchActionFoo',
```
every time to stop with variable calls.

mpoiriert replied:
Yes the controller is not just the class but the full path with function like you would define a specific path to a controller in a routing configuration. If you just specify the class in imply __invoke (I think).

I was keeping the finalAction action logic to be backward compatible. But I could try your way to implement it.

I was thinking about deprecated it as well. But it could be a nice short cut the same way we don't need to specify the type when we create a form because it get auto detected...

VincentLanglet commented on src/Controller/CRUDController.php:
```diff
@@ -524,7 +523,60 @@ public function batchAction(Request $request): Response
+            throw new \LogicException(sprintf('Batch action `%s` cannot be executed. You must call `%s::canBatchActionBeExecuted` first.', $action, static::class));
+        }
+
+        $controller = $this->admin->getBatchActions()[$action]['controller'];
```
You need `?? null`.

VincentLanglet commented on src/Controller/CRUDController.php:
```diff
@@ -524,7 +523,60 @@ public function batchAction(Request $request): Response
+
+    final protected function getBatchActionExecutable(string $action): callable
+    {
+        if (!$this->canBatchActionBeExecuted($action)) {
+            throw new \LogicException(sprintf('Batch action `%s` cannot be executed. You must call `%s::canBatchActionBeExecuted` first.', $action, static::class));
+        }
```
This means the canBatchActionBeExecuted method is executed twice... :s

mpoiriert replied:
Yes it is. I was going to put the method private, if we do we could not call it again since the flow we have will make sure it cannot happen. If we keep it protected I think it's better to have the check. I'll put it private for now and wait if we have a need to put in protected at some point.

VincentLanglet commented on src/Controller/CRUDController.php:
```diff
@@ -439,12 +442,15 @@ public function batchAction(Request $request): Response
+
+        if (!$this->canBatchActionBeExecuted($action)) {
+            $finalAction = sprintf('batchAction%s', $camelizedAction);
+            throw new \RuntimeException(sprintf('A `%s::%s` method must be callable or create add a `controller` configuration', static::class, $finalAction));
```
There are some error in the exception message.

mpoiriert committed:
Adjust base on code review

mpoiriert commented:
@VincentLanglet I have addressed your comment. I also have deprecated the isRelevant hook.

I have tested the delete action with the new naming convention and it's fine.

We must decide if we want to keep the naming convention with the action (as a auto configuration) or not. If not I will deprecate it.

We could also move the logic in the admin AbstractAdmin::getBatchActions so the CRUDController have less logic and the admin would be responsible of the full configuration.

VincentLanglet commented on src/Controller/CRUDController.php:
```diff
@@ -524,7 +530,48 @@ public function batchAction(Request $request): Response
+                static::class,
+                sprintf('batchAction%s',  InflectorFactory::create()->build()->classify($action))
+            );
+        }
```
```suggestion
        $controller = $batchActions[$action]['controller'] ?? sprintf(
            '%s::%s',
            static::class,
            sprintf('batchAction%s',  InflectorFactory::create()->build()->classify($action))
       );
```

VincentLanglet reviewed (commented):
There is some doc about the isRelevant part that we can remove (or say it's deprecated).

mpoiriert committed:
Fix resolving batch action controller

Co-authored-by: Vincent Langlet <VincentLanglet@users.noreply.github.com>

mpoiriert commented:
When we are ok with the code I will pass trough documentation and also the automation test. Noted for the mention of "isRelevant".

VincentLanglet commented:
> When we are ok with the code I will pass trough documentation and also the automation test. Noted for the mention of "isRelevant".

The implementation seems ok to me. let's wait for another look from @jordisala1991  or @franmomu

mpoiriert committed:
fix related to PHP-CS-Fixer

mpoiriert commented:
@VincentLanglet I am trying to fix the linter error.

This one: 

Method Sonata\AdminBundle\Controller\CRUDController::getBatchActionController() has no return type specified. 

But controller in symfony can be a lot of different thing (string, array, callable, etc.) Is there a way to ignore it ? Or I need to put a @return string|array|callable.

I will not be able to specify the return type on the method since php 7 doesn't support mixed or the | operator.

VincentLanglet commented:
> @VincentLanglet I am trying to fix the linter error.
> 
> This one:
> 
> Method Sonata\AdminBundle\Controller\CRUDController::getBatchActionController() has no return type specified.
> 
> But controller in symfony can be a lot of different thing (string, array, callable, etc.) Is there a way to ignore it ? Or I need to put a @return string|array|callable.
> 
> I will not be able to specify the return type on the method since php 7 doesn't support mixed or the | operator.

`@return` annotation is ok to fix the error.

If you use `array`, you'll have to say what's in your array.
Like `array{string, string}` or `array<string, mixed>` or `array{foo: int}`, etc

franmomu commented on src/Controller/CRUDController.php:
```diff
@@ -1478,6 +1483,48 @@ final protected function checkParentChildAssociation(Request $request, object $o
+
+        return function (ProxyQueryInterface $query, Request $request) use ($controller) {
+            $request->attributes->set('_controller', $controller);
+            $request->attributes->set('query', $query);
```
I find this a bit tricky, with this we are storing the whole query in session to be able to pass it to a controller.

`batchAction` needs to be an action because it's called directly from the view, but in this case we are processing the data sent, creating the query and making some other checks, so to me it seems more like a service implementing an interface receiving explicitly `ProxyQueryInterface $query` and `Request $request` rather than a controller that is not going to be called "as a controller".

On the other hand, if this new controller was able to replace the whole `batchAction`, I think that would be a good move (reusing somehow the `batchAction` code and would be a step forward to split/remove the `CRUDController`).

VincentLanglet replied:
> I find this a bit tricky, with this we are storing the whole query in session to be able to pass it to a controller.
> 
> `batchAction` needs to be an action because it's called directly from the view, but in this case we are processing the data sent, creating the query and making some other checks, so to me it seems more like a service implementing an interface receiving explicitly `ProxyQueryInterface $query` and `Request $request` rather than a controller that is not going to be called "as a controller".

Yeah, that's why I first proposed to use a service like this: https://github.com/sonata-project/SonataAdminBundle/pull/7819#discussion_r873200786
but it was considered as too complex. Do you have a better solution ?

> On the other hand, if this new controller was able to replace the whole `batchAction`, I think that would be a good move (reusing somehow the `batchAction` code and would be a step forward to split/remove the `CRUDController`).

How would you do this ?

mpoiriert replied:
> I find this a bit tricky, with this we are storing the whole query in session to be able to pass it to a controller.

Not sure exactly what you mean by "session" the request attribute is not a session. Unless you didn't mean the php session.

It's just a simple solution to injection the parameter in the controller since Symfony check if the parameter is available in the request.

It's use often in symfony you just have to look for **attributes->set(** in vendor/symfony.

mpoiriert replied:
> rather than a controller that is not going to be called "as a controller".

Not sure I agree, a controller jobs is to take request validate the data and return the proper response. Which is exactly what it does.

From my point of view I consider that if the code don't need to be reuse (could also include automation test purpose) it doesn't need to be in a dedicated service.

In Symfony you can bind a service method to a route or inject a controller in another service/controller, so having it define as a service or controller is not that crucial. Rules of thumbs, the less amount lines (of code, documentation, configuration) needed to do something is (almost) always the best way. At the end it need to be as easy as possible for the end user to implement.

If you want to remove the CRUDController at some point we could move the code the batchAction to a single file controller/action.

VincentLanglet replied:
> From my point of view I consider that if the code don't need to be reuse (could also include automation test purpose) it doesn't need to be in a dedicated service.

If the same batch actions is used for differents admins, you'll reuse your batch action ; so the code is kinda reuse then.

> rather than a controller that is not going to be called "as a controller".

@franmomu Would you prefer an invocable BatchAction ?

```
interface BatchActionInterface
{
    __invoke(ProxyQueryInterface $proxyQuery, Request $request);
}
```

But then we'll need a way to access to the batch actions
```
($this->container->get(...))($proxyQuery, $request);
```
Not sure if there is a clean way to do this.

franmomu replied:
Sorry about the attributes an session, I got confused, what I meant is that this parameter bag I think is usually used for event listeners or services to add/retrieve custom information, but in this case it would be the user the one to have to call `$request->attributes->get('query')` which I don't find it simple or intuitive.

That's why I prefer explicit parameters like @VincentLanglet said, which I find it cleaner, but I can't think of right know of a good solution to register/fetch these actions easily by admin.

Maybe the ParamConverter/ArgumentResolver could be a solution so the user only has to define the action like:

```php
public function batchClone(BatchAction $batchAction): Response
{
    $admin = $batchAction->getAdmin();
    $query = $batchAction->getQuery();
    // whatever is needed
}
```

mpoiriert replied:
> If the same batch actions is used for differents admins, you'll reuse your batch action ; so the code is kinda reuse then

Reuse in the sens of called from different flow, meaning "copy pasting" code causing duplication. This is not the case.

mpoiriert replied:
> call `$request->attributes->get('query')` which I don't find it simple or intuitive.

Not really since by putting it in the attributes it will get injected in the controller method like I did put in my example. No need to manipulate anything.

> Maybe the ParamConverter/ArgumentResolver could be a solution so the user only has to define the action like:

I would suggest spiting this.

```PHP
public function batchClone(ProxyQueryInterface $query, AdminInterface $admin): Response
{
    // whatever is needed
}
```

The $query is already done the way I have implement it, doing a param converter would not add anything.
The $admin could be done separately via a ParameterConverter that use the AdminFectcherInterface, that way we could reuse it every where (improving what we have to get the Admin)

mpoiriert replied:
> interface BatchActionInterface
> {
>     __invoke(ProxyQueryInterface $proxyQuery, Request $request);
> }

This is forcing user to follow stick implementation where it's not needed. The beautify of the symfony controller is the flexibility.

Someone could do something like this

```PHP
class CloningController
{
     __invoke(ProxyQueryInterface $proxyQuery, AdminIterface $admin, ClonerInterface $cloner): Response
    {
       $cloner->clone($admin, $proxyQuery);
       return $this->redirectToList($admin);
    }
}

```

mpoiriert replied:
Just to give a real example of what it would look like after what I suggest for a real use case that I have for one of my client that rely on sonata media bundle that now have a final controller and a final Media admin:

```PHP
class MediaExtension extends AbstractAdminExtension
{
    public function configureBatchActions(AdminInterface $admin, array $actions): array
    {
        if ($admin instanceof BaseMediaAdmin) {
            $actions['allow_compression'] = [
                'label' => 'Allow compression',
                'ask_confirmation' => false,
                'controller' => MediaController::class . '::batchAllowCompressionAction',
            ];
        }
        
        return $actions;
    }
}

class MediaController extends AbstractController
{
    public function batchAllowCompressionAction(
        EntityManagerInterface $entityManager,
        ProxyQueryInterface $query,
        AdminInterface $admin
    ): Response {
        $medias = $query->execute();
        $compressionRequested = 0;

        try {
            /** @var Media $media */
            foreach ($medias as $media) {
                if ($media->setAllowCompression(true)) {
                    ++$compressionRequested;
                }
            }

            if ($compressionRequested) {
                $entityManager->flush();
            }

            $this->addFlash(
                'sonata_flash_success',
                sprintf('Media proceed - %s, compression requested - %s.', \count($medias), $compressionRequested)
            );
        } catch (Throwable $e) {
            $this->addFlash('sonata_flash_error', 'An error occur when allowing compression: ' . $e->getMessage());
        } finally {
            return new RedirectResponse(
                $admin->generateUrl('list', ['filter' => $admin->getFilterParameters()])
            );
        }
    }
}
```

VincentLanglet replied:
This seems ok to me

franmomu replied:
yep, that seems fine

mpoiriert commented:
The solution I see that would be simple and matching Symfony way is doing a
param converter to the ProxyQueryInterface.

So the person will only need to declare the controller and we will inject
the query in it.

The same way Doctrine Param Converter return a 404 we could maybe find a
wait to redirect to the list view if the query is empty.

On Tue., May 17, 2022, 3:18 a.m. Vincent Langlet, ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In src/Controller/CRUDController.php
> <https://github.com/sonata-project/SonataAdminBundle/pull/7819#discussion_r874452837>
> :
>
> > +        $controller = $batchActions[$action]['controller'] ?? sprintf(
> +            '%s::%s',
> +            static::class,
> +            sprintf('batchAction%s', InflectorFactory::create()->build()->classify($action))
> +        );
> +
> +        return $controller;
> +    }
> +
> +    private function getBatchActionExecutable(string $action): callable
> +    {
> +        $controller = $this->getBatchActionController($action);
> +
> +        return function (ProxyQueryInterface $query, Request $request) use ($controller) {
> +            $request->attributes->set('_controller', $controller);
> +            $request->attributes->set('query', $query);
>
> I find this a bit tricky, with this we are storing the whole query in
> session to be able to pass it to a controller.
>
> batchAction needs to be an action because it's called directly from the
> view, but in this case we are processing the data sent, creating the query
> and making some other checks, so to me it seems more like a service
> implementing an interface receiving explicitly ProxyQueryInterface $query
> and Request $request rather than a controller that is not going to be
> called "as a controller".
>
> Yeah, that's why I first proposed to use a service like this: #7819
> (comment)
> <https://github.com/sonata-project/SonataAdminBundle/pull/7819#discussion_r873200786>
> but it was considered as too complex. Do you have a better solution ?
>
> On the other hand, if this new controller was able to replace the whole
> batchAction, I think that would be a good move (reusing somehow the
> batchAction code and would be a step forward to split/remove the
> CRUDController).
>
> How would you do this ?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/sonata-project/SonataAdminBundle/pull/7819#discussion_r874452837>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AA73OAB6IUXWKNG3R5KVW23VKNB4TANCNFSM5V7KMPOA>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

mpoiriert commented:
I'll continue working on it and fix the automation test. I might do a
specific PR for the Admin interface param converter.

I'll try to do this in the next couple of weeks.

On Thu., May 19, 2022, 4:27 p.m. Fran Moreno, ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In src/Controller/CRUDController.php
> <https://github.com/sonata-project/SonataAdminBundle/pull/7819#discussion_r877511852>
> :
>
> > +        $controller = $batchActions[$action]['controller'] ?? sprintf(
> +            '%s::%s',
> +            static::class,
> +            sprintf('batchAction%s', InflectorFactory::create()->build()->classify($action))
> +        );
> +
> +        return $controller;
> +    }
> +
> +    private function getBatchActionExecutable(string $action): callable
> +    {
> +        $controller = $this->getBatchActionController($action);
> +
> +        return function (ProxyQueryInterface $query, Request $request) use ($controller) {
> +            $request->attributes->set('_controller', $controller);
> +            $request->attributes->set('query', $query);
>
> yep, that seems fine
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/sonata-project/SonataAdminBundle/pull/7819#discussion_r877511852>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AA73OAAGRF64LOC6W2Q2QHTVK2P4RANCNFSM5V7KMPOA>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

VincentLanglet commented:
Don't we already have the paramConverter thanks to https://github.com/sonata-project/SonataAdminBundle/blob/4.x/src/ArgumentResolver/AdminValueResolver.php ?

mpoiriert commented:
Seems like like it. So I just need to fix the automation test and adjust
the documentation then ?
On Thu., May 19, 2022, 4:43 p.m. Vincent Langlet, ***@***.***>
wrote:

> Don't we already have the paramConverter thanks to
> https://github.com/sonata-project/SonataAdminBundle/blob/4.x/src/ArgumentResolver/AdminValueResolver.php
> ?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/sonata-project/SonataAdminBundle/pull/7819#issuecomment-1132190311>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AA73OAGLZSGFTTDO5IDAZ7TVK2RXBANCNFSM5V7KMPOA>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

VincentLanglet commented:
> Seems like like it. So I just need to fix the automation test and adjust the documentation then ?
> […](#)
> On Thu., May 19, 2022, 4:43 p.m. Vincent Langlet, ***@***.***> wrote: Don't we already have the paramConverter thanks to https://github.com/sonata-project/SonataAdminBundle/blob/4.x/src/ArgumentResolver/AdminValueResolver.php ? — Reply to this email directly, view it on GitHub <[#7819 (comment)](https://github.com/sonata-project/SonataAdminBundle/pull/7819#issuecomment-1132190311)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AA73OAGLZSGFTTDO5IDAZ7TVK2RXBANCNFSM5V7KMPOA> . You are receiving this because you authored the thread.Message ID: ***@***.***>

Thinks so. And adding a functional test with the way to use it will be a nice proof that it works as expected :)

mpoiriert commented:
@VincentLanglet I am trying to run automation test on my local computer. I get a bunch of error like those (even on the master branch).


1) Sonata\AdminBundle\Tests\Controller\CRUDControllerTest::testEditAction
assert(): assert(null !== $objectId) failed

/home/wwwroot/src/Controller/CRUDController.php:290
/home/wwwroot/tests/Controller/CRUDControllerTest.php:1459
phpvfscomposer:///home/wwwroot/vendor/phpunit/phpunit/phpunit:97

Not sure why it's happening.

Other than a composer install and make test is there anything else I need to do ?

VincentLanglet commented:
Will be fixed by https://github.com/sonata-project/SonataAdminBundle/pull/7831

mpoiriert committed:
fix tests

VincentLanglet commented:
https://github.com/sonata-project/SonataAdminBundle/pull/7831 is merged if you want to rebase @mpoiriert

mpoiriert committed:
Documentation

mpoiriert commented:
@VincentLanglet I have rebase fix last linter and update documentation. Hopefully we are almost there...

mpoiriert committed:
Fix test php-lint

franmomu commented on src/Controller/CRUDController.php:
```diff
@@ -1478,6 +1485,36 @@ final protected function checkParentChildAssociation(Request $request, object $o
+            sprintf('batchAction%s', InflectorFactory::create()->build()->classify($action))
+        );
+
+        // This will throw an exception when called so we know if it's possible or not to call the controller.
+        $exists = false !== $this->container
+            ->get('controller_resolver')
+            ->getController(new Request([], [], ['_controller' => $controller]));
+
+        if (!$exists) {
+            throw new \RuntimeException(sprintf('Controller for action `%s` cannot be resolved', $action));
+        }
```
One question, is this needed? If the `controller` is not a valid one I guess the `$this->container->get('http_kernel')->handle()` would also throw an exception.

mpoiriert replied:
@franmomu yes you are right but I am relying on the exception here for the test prior to execution of the query, the ask_confirmation, etc to tell the user the action is not configured properly. If we do it on a try and catch of the execution we would not easily know which exception to throw since it could be a exception from the execution. Once we move to version 5 we could just not catch the exception and have the default Symfony error stating the controller is not found and not bother about explaining how to configure the batch action controller.

mpoiriert committed:
fix documentation linter

mpoiriert commented:
I am working on symfony 6 issues. I am wondering why this one is consider a failed ?

https://github.com/sonata-project/SonataAdminBundle/runs/6573865897?check_suite_focus=true

mpoiriert committed:
fix tests

VincentLanglet commented:
> I am working on symfony 6 issues. I am wondering why this one is consider a failed ?
> 
> https://github.com/sonata-project/SonataAdminBundle/runs/6573865897?check_suite_focus=true

Because of the `Remaining self deprecation notices`.

If something is deprecated/throw a deprecation, you have to add the `@group legacy` on the test.
```
/** 
 * @group legacy
 *
 * NEXT_MAJOR: Remove this test
 */
```

mpoiriert committed:
fix doc

mpoiriert commented:
We are almost there sorry for the back and forth a lot of thing need to be checked and I cannot run all the linter locally. Next time I will not need a maintainer to run the workflow, I should be able to handle things on my own.

VincentLanglet commented:
/home/runner/work/SonataAdminBundle/SonataAdminBundle/docs/reference/batch_actions.rst:220:Title underline too short.

jordisala1991 commented on docs/reference/batch_actions.rst:
```diff
@@ -149,9 +216,11 @@ a radio button to choose the target object.
 
-(Optional) Overriding the default relevancy check function
+(Optional|Deprecated) Overriding the default relevancy check function
 ----------------------------------------------------------
```
```suggestion
---------------------------------------------------------------------
```

mpoiriert replied:
I fixed it

mpoiriert commented:
Sorry the file wasn't push before the last workflow got triggered...

jordisala1991 commented on docs/reference/batch_actions.rst:
```diff
@@ -46,6 +48,71 @@ merges them onto a single target item. It should only be available when two cond
+
+            $target = $modelManager->find($admin->getClass(), $request->get('targetId'));
+
+            if ($target === null){
```
```suggestion
            if ($target === null) {
```

jordisala1991 commented on docs/reference/batch_actions.rst:
```diff
@@ -37,6 +37,8 @@ merges them onto a single target item. It should only be available when two cond
         ) {
             $actions['merge'] = [
                 'ask_confirmation' => true
```
```suggestion
                'ask_confirmation' => true,
```

jordisala1991 commented on docs/reference/batch_actions.rst:
```diff
@@ -37,6 +37,8 @@ merges them onto a single target item. It should only be available when two cond
         ) {
             $actions['merge'] = [
                 'ask_confirmation' => true
+                'controller' => 'app.controller.merge::batchMergeAction'
```
```suggestion
                'controller' => 'app.controller.merge::batchMergeAction',
```

mpoiriert committed:
Update docs/reference/batch_actions.rst

Co-authored-by: Jordi Sala Morales <jordism91@gmail.com>

mpoiriert committed:
Update docs/reference/batch_actions.rst

Co-authored-by: Jordi Sala Morales <jordism91@gmail.com>

mpoiriert committed:
Update docs/reference/batch_actions.rst

Co-authored-by: Jordi Sala Morales <jordism91@gmail.com>

mpoiriert commented:
Is there anything missing for this PR ?

VincentLanglet commented on src/Controller/CRUDController.php:
```diff
@@ -439,15 +442,26 @@ public function batchAction(Request $request): Response
         $isRelevantAction = sprintf('batchAction%sIsRelevant', $camelizedAction);
 
         if (method_exists($this, $isRelevantAction)) {
+            // TODO: Remove if above in sonata-project/admin-bundle 5.0
```
```suggestion
            // NEXT_MAJOR: Remove if above in sonata-project/admin-bundle 5.0
```

VincentLanglet commented:
> Is there anything missing for this PR ?

It's ok for me ; I just have a small request change

mpoiriert committed:
Change TODO for NEXT_MAJOR

Co-authored-by: Vincent Langlet <VincentLanglet@users.noreply.github.com>

jordisala1991 commented on docs/reference/batch_actions.rst:
```diff
@@ -46,6 +48,71 @@ merges them onto a single target item. It should only be available when two cond
+                );
+            }
+
+            $selectedModels = $selectedModelQuery->execute();
```
```suggestion
            $selectedModels = $query->execute();
```

Is this suposed to be $query ? I do not find $selectedModelQuery anywhere

mpoiriert committed:
Fix documentation example $query variable

Co-authored-by: Jordi Sala Morales <jordism91@gmail.com>

franmomu commented:
Some comments, I was thinking if we can use real instances instead of mocks, I think that would simplify tests (removing callback assertions). 

For `CRUDControllerTest`, I guess we can use `ContainerControllerResolver` which would receive `$this->controller` and `HttpKernel` with a stub of an `EventDispatcherInterface` and `$this->controllerResolver`.

That way instead of add expectations we could register controllers in the container and execute those controllers (theoretically).

I can try to see if it works and if it's simpler later today.

What I think it's missing is a functional test adding a batch action from another controller.

mpoiriert commented:
@franmomu I agree with you about the mock. I normally do the less amount of mock in my project and do integration testing first. But I think refactoring the CRUDControllerTest is out of scope.

Doing an integration test would be nice but we are pretty much testing if 
Symfony is working as expected since we are just doing what the mock are doing. 

I don't think we should have both (mock and integration). It's a bit redundant.

I think that in Sonata 5 the batch action should be it's own controller (not in CRUDController) as well as the batch delete. Maybe in those we could do a integration test by having a App/Kernel for the test, loading proper bundle/configuration and test if it's working.

mpoiriert commented:
@VincentLanglet Is there something missing here ?

mpoiriert commented:
@VincentLanglet Is there something missing here ?

VincentLanglet commented:
> @VincentLanglet Is there something missing here ?

Another approval.

cc @franmomu or @jordisala1991

Hanmac commented:
@mpoiriert i don't know if that info can be accessed somewhere else,
but assuming you have 2 Admin using the same batch action (calling the same controller, and assuming that this controller isn't a CRUD one)
is there a way to give it the Admin who called the batch action as parameter?

so while inside the batch action you can see which admin called it?

VincentLanglet commented:
I would say you can always access to the admin thanks to the admin fetcher

mpoiriert commented:
@Hanmac As you can see in the updated documentation here: https://github.com/sonata-project/SonataAdminBundle/pull/7819/files#diff-4a755e8509419b197a92e0d368c1b4989dbc4863336df9b6c0b04e2e39e5cd39

Hanmac commented:
thanks you two, so i just need to wait for this to get merged

(and then maybe find a better way to define a multi step batch action somehow)

jordisala1991 reviewed (commented):
This should probably require an upgrade note if there is something that will be removed on next major

mpoiriert committed:
Add batch action is relevent upgrade documentation

mpoiriert commented:
I have rebase 4.x and added a upgrade documentation for the Deprecate `batchAction%sIsRelevant` hook

VincentLanglet commented:
Can we get another approval (@sonata-project/contributors @jordisala1991 @franmomu)
